### PR TITLE
`exec` refactoring

### LINKED ISSUE #942: Docksal Support for Alpine Images
Right now fin seems to have a hard dependency on `bash` for many operations in the containers to function.  This means that when running containers (like Alpine based images) you will run into many different issues with various fin commands. `fin bash` and `fin exec` being the 2 most obvious commands.  There are work arounds for some of these things by falling back to docker-compose, but that limits the benefits of fin greatly. `fin dc run <container> /bin/ash` for example will get you to a shell in an Alpine based image.

`node:10.15.0-alpine` was the specific image I was testing with from https://hub.docker.com/_/node/ for reference.

### PR DESCRIPTION
**Pull request by achekulaev**
`exec` and `bash` will use sh by default and only use bash for `cli` and `db` (complex db functions input requires `bash` as well).

Closes #942 
Relates to #924

### CODE REVIEW

File: bin/fin [hunk 1]
```diff
@@ -4711,66 +4719,75 @@ _exec ()
...
+	else
+		container_name="cli"
 	fi
 
-	CONTAINER_NAME=${CONTAINER_NAME:-cli}
-	container_id=$(get_container_id "$CONTAINER_NAME")
+	container_id=$(get_container_id "$container_name")
+
+	# Use sh as default shell
+	local SHELL_INTERACTIVE="sh -ilc"
```
Comment (lmakarov) [inline bin/fin]: Use lowercase variable names for local variables, to keep things consistent

File: bin/fin [hunk 2]
```diff
@@ -4711,66 +4719,75 @@ _exec ()
...
+	local SHELL_NONINTERACTIVE="sh -lc"
+
+	# For cli use bash
+	if [[ "$container_name" == "cli" ]] || [[ "$container_name" == "db" ]]; then
+		SHELL_INTERACTIVE="bash -ilc"
```
Comment (lmakarov) [inline bin/fin]: Use lowercase variable names for local variables, to keep things consistent

File: bin/fin [hunk 3]
```diff
@@ -4711,66 +4719,75 @@ _exec ()
...
+	# Add space if cmd is not empty
+	[[ $cmd != "" ]] && cmd="$cmd "
 	# Escape spaces that are "spaces" and not parameter delimiters (i.e., param1 param2\ with\ spaces param3)
 	if [[ $2 != "" ]]; then
-		cmd="$cmd "$(printf " %q" "$@")
+		cmd="$cmd"$(printf " %q" "$@")
 	# Do not escape spaces if there is only one parameter (e.g., fin run "ls -la | grep txt")
 	else
-		cmd="$cmd $@"
+		cmd="$cmd""$@"
```
Comment (lmakarov) [inline bin/fin]: `cmd="${cmd}${@}"` would be a cleaner option

File: docs/content/fin/fin-help.md [hunk 4]
```diff
@@ -1,5 +1,5 @@
 ---
-title: "All commands reference"
+title: "fin help"
```
Comment (lmakarov) [inline docs/content/fin/fin-help.md]: Keep the title "All commands reference"

File: bin/fin [hunk 5]
```diff
@@ -4732,7 +4732,7 @@ _exec ()
 	local SHELL_NONINTERACTIVE="sh -lc"
 
 	# For cli use bash
```
Comment (lmakarov) [inline bin/fin]: Add a comment around why bash has to be used for `db`

File: bin/fin [hunk 6]
```diff
@@ -4679,20 +4681,26 @@ image_registry_list ()
...
 	if ! is_tty; then
-		echo "not a tty"
+		echo-error "tty is required to start an interactive shell"
 		return 1
 	fi
 
-	# Pass container name to _run
-	CONTAINER_NAME=$1 _exec bash -il
+	# Use sh by default
+	if [[ "$1" != "" ]]; then
```
[the remark this answers is not in the document]
Comment (achekulaev) [reply bin/fin]: It is bash function it does not accept --in parameter

File: bin/fin (hunk 1 above)
Comment (achekulaev) [reply bin/fin]: It is a constant. Hence uppercase

File: docs/content/fin/fin-help.md (hunk 4 above)
Comment (achekulaev) [reply docs/content/fin/fin-help.md]: It is the title that you set in help generation script. I manually changed it in another commit but the script replaced it again.

File: bin/fin (hunk 5 above)
Comment (achekulaev) [reply bin/fin]: done

File: bin/fin (hunk 3 above)
Comment (achekulaev) [reply bin/fin]: done

File: docs/content/fin/fin-help.md (hunk 4 above)
Comment (lmakarov) [reply docs/content/fin/fin-help.md]: Right. Let's then leave it as set in `scripts/fin-help-generate.sh` - "fin help"

**Pull request merged**